src/ott/solvers/nn/neuraldual.py

```diff
@@ -9,76 +9,131 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""A Jax implementation of the ICNN based Kantorovich dual."""
+"""A Jax implementation of the neural-based Kantorovich dual."""
 
 import warnings
-from typing import Dict, Iterable, List, Literal, Optional, Tuple, Union
+from typing import Callable, Dict, Iterable, List, Literal, Optional, Tuple, Union
 
-import flax.linen as nn
 import jax
 import jax.numpy as jnp
 import optax
 from flax import core
+from flax.core import frozen_dict
 
 from ott.geometry import costs
 from ott.problems.linear import potentials
-from ott.solvers.nn import icnn
+from ott.solvers.nn import conjugate_solvers, models
 
-__all__ = ["NeuralDualSolver"]
+__all__ = ["W2NeuralDual"]
 
 Train_t = Dict[Literal["train_logs", "valid_logs"], Dict[str, List[float]]]
-
-
-class NeuralDualSolver:
-  r"""Solver of the ICNN-based Kantorovich dual.
-
-  Learn the optimal transport between two distributions, denoted source
-  and target, respectively. This is achieved by parameterizing the two
-  Kantorovich potentials, `g` and `f`, by two input convex neural networks.
-  :math:`\nabla g` hereby transports source to target cells, and
-  :math:`\nabla f` target to source cells.
-
-  Original algorithm is described in :cite:`makkuva:20`.
+Callback_t = Callable[[int, potentials.DualPotentials], None]
+Conj_t = Optional[conjugate_solvers.FenchelConjugateSolver]
+
+
+class W2NeuralDual:
+  r"""Solver for the Wasserstein-2 Kantorovich dual between Euclidean spaces.
+
+  Learn the Wasserstein-2 optimal transport between two measures
+  :math:`\alpha` and :math:`\beta` in
+  :math:`n`-dimensional Euclidean space,
+  denoted source and target, respectively.
+  This is achieved by parameterizing a Kantorovich potential
+  :math:`f_\theta: \mathbb{R}^n\rightarrow\mathbb{R}`
+  associated with the :math:`\alpha` measure with
+  an :class:`~ott.solvers.nn.models.ICNN`,
+  :class:`~ott.solvers.nn.models.MLP`,
+  or other
```

Comment:
Pet peeve: don't like this almost empty line (same for the `where` below).

Reply:
Joined them (although I like having long paths on their own line since it makes quickly editing them slightly easier)